experiments/AtmosGCM/moist-baroclinic-wave.jl

```diff
@@ -0,0 +1,223 @@
+    temp_profile_ref = IsothermalProfile(param_set, FT(300.0))
+    ref_state = HydrostaticState(temp_profile_ref)
+
+    domain_height::FT = 44e3 # distance between surface and top of atmosphere (m)
+
+    # Set up the atmosphere model
+    exp_name = "MoistBaroclinicWave"
+
+    model = AtmosModel{FT}(
+        AtmosGCMConfigType,
+        param_set;
+        ref_state = ref_state,
+        turbulence = ConstantViscosityWithDivergence(FT(0.0)),
+        moisture = DryModel(),
+        source = (Gravity(), Coriolis(),),
+        init_state_conservative = init_moist_baroclinic_wave!,
+    )
+
+    config = ClimateMachine.AtmosGCMConfiguration(
+        exp_name,
+        poly_order,
+        resolution,
+        domain_height,
+        param_set,
+        init_moist_baroclinic_wave!;
+        model = model,
+    )
+
+    return config
+end
+
+function config_diagnostics(FT, driver_config)
+    interval = "40000steps" # chosen to allow a single diagnostics collection
+
+    _planet_radius = FT(planet_radius(param_set))
+
+    info = driver_config.config_info
+    boundaries = [
+        FT(-90.0) FT(-180.0) _planet_radius
+        FT(90.0) FT(180.0) FT(_planet_radius + info.domain_height)
+    ]
+    resolution = (FT(10), FT(10), FT(1000)) # in (deg, deg, m)
+    interpol = ClimateMachine.InterpolationConfiguration(
+        driver_config,
+        boundaries,
+        resolution,
+    )
+
+    dgngrp = setup_atmos_default_diagnostics(
+        AtmosGCMConfigType(),
+        interval,
+        driver_config.name,
+        interpol = interpol,
+    )
+
+    return ClimateMachine.DiagnosticsConfiguration([dgngrp])
+end
+
+function main()
+    # Driver configuration parameters
+    FT = Float64                             # floating type precision
+    poly_order = 5                           # discontinuous Galerkin polynomial order
+    n_horz = 10                              # horizontal element number
+    n_vert = 6                               # vertical element number
+    n_days::FT = 30 
+    timestart::FT = 0                        # start time (s)
+    timeend::FT = n_days * days(param_set)   # end time (s)
+
+    # Set up driver configuration
+    driver_config = config_moist_baroclinic_wave(FT, poly_order, (n_horz, n_vert))
+
+    # Set up experiment
+    CFL::FT = 0.2
+    solver_config = ClimateMachine.SolverConfiguration(
+        timestart,
+        timeend,
+        driver_config,
+        Courant_number = CFL,
+        CFL_direction = HorizontalDirection(),
+        diffdir = HorizontalDirection(),
+    )
+
+    # Set up diagnostics
+    dgn_config = config_diagnostics(FT, driver_config)
+
+    # Set up user-defined callbacks
+    filterorder = 64
+    filter = ExponentialFilter(solver_config.dg.grid, 0, filterorder)
+    cbfilter = GenericCallbacks.EveryXSimulationSteps(1) do
+        @views begin
+          solver_config.Q.data[:, 1, :] .-= solver_config.dg.state_auxiliary.data[:, 8, :]
```

Comment:
should be able to use vars aware MPIStateArrays here to identify state indices with their `@vars names`

Reply:
This should hopefully get cleaned up once #1200 and #1182 land